File: nodejs/test/test_encoder.js
import assert from 'node:assert/strict';
import { Buffer } from 'node:buffer';
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { describe, it } from 'node:test';

import {
  adaptClientType,
  encodeValue,
  formatResultRow,
  formatValue,
  literalFormatConfig,
  simpleFormatConfig,
  boolValue,
  isNullValue,
  listValues,
  numberValue,
  stringValue,
  typeCode,
  valueKind,
  arrayElementType,
  fieldType,
  structFields,
  TypeCode,
} from '../lib/index.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const CONFORMANCE_PATH = path.resolve(__dirname, '../../testdata/conformance.json');

/** @returns {import('../../testdata/conformance.json')} */
function loadConformance() {
  return JSON.parse(readFileSync(CONFORMANCE_PATH, 'utf8'));
}

/**
 * @param {unknown} a
 * @param {unknown} b
 */
function wireEqual(a, b) {
  assert.deepEqual(normalizeWire(a), normalizeWire(b));
}

/**
 * @param {unknown} value
 */
function normalizeWire(value) {
  if (value === null || isNullValue(value)) {
    return null;
  }
  const kind = valueKind(value);
  if (kind === 'bool') {
    return boolValue(value);
  }
  if (kind === 'number') {
    return numberValue(value);
  }
  if (kind === 'string') {
    return stringValue(value);
  }
  if (kind === 'list') {
    return listValues(value).map((elem) => normalizeWire(elem));
  }
  return value;
}

/**
 * Decode conformance wire value to a native value for encode round-trips.
 * @param {unknown} typ
 * @param {unknown} value
 */
function wireToNative(typ, value) {
  if (value === null || isNullValue(value)) {
    return null;
  }
  const code = typeCode(typ);
  switch (code) {
    case TypeCode.BOOL:
      return boolValue(value);
    case TypeCode.INT64:
    case TypeCode.ENUM:
      return stringValue(value);
    case TypeCode.FLOAT32:
    case TypeCode.FLOAT64: {
      const kind = valueKind(value);
      if (kind === 'string') {
        const s = stringValue(value);
        if (s === 'NaN') return Number.NaN;
        if (s === 'Infinity') return Infinity;
        if (s === '-Infinity') return -Infinity;
      }
      return numberValue(value);
    }
    case TypeCode.BYTES:
    case TypeCode.PROTO:
      return Buffer.from(stringValue(value), 'base64');
    case TypeCode.STRING:
    case TypeCode.TIMESTAMP:
    case TypeCode.DATE:
    case TypeCode.NUMERIC:
    case TypeCode.INTERVAL:
    case TypeCode.UUID:
    case TypeCode.JSON:
      return stringValue(value);
    case TypeCode.ARRAY: {
      const elemType = arrayElementType(typ);
      return listValues(value).map((elem) => wireToNative(elemType, elem));
    }
    case TypeCode.STRUCT: {
      const fields = structFields(typ);
      return listValues(value).map((elem, i) => wireToNative(fieldType(fields[i]), elem));
    }
    default:
      return value;
  }
}

const conformance = loadConformance();

describe('encodeValue round-trip', () => {
  for (const testCase of conformance.value_cases) {
    it(testCase.name, () => {
      const native = wireToNative(testCase.type, testCase.value);
      const encoded = encodeValue(testCase.type, native);
      wireEqual(encoded, testCase.value);
      assert.equal(
        formatValue(testCase.type, encoded, simpleFormatConfig()),
        testCase.expected.simple,
        `format after encode for ${testCase.name}`,
      );
    });
  }
});

describe('encodeValue struct hash', () => {
  it('encodes named struct from object', () => {
    const typ = {
      code: 'STRUCT',
      structType: {
        fields: [
          { name: 'n', type: { code: 'INT64' } },
          { name: 's', type: { code: 'STRING' } },
        ],
      },
    };
    const encoded = encodeValue(typ, { n: '1', s: 'foo' });
    wireEqual(encoded, ['1', 'foo']);
    assert.equal(
      formatValue(typ, encoded, literalFormatConfig()),
      'STRUCT<n INT64, s STRING>(1, "foo")',
    );
  });
});

describe('adaptClientType', () => {
  it('adapts camelCase client shapes', () => {
    const client = {
      code: 'ARRAY',
      arrayElementType: { code: 'INT64' },
    };
    assert.deepEqual(adaptClientType(client), {
      code: 'ARRAY',
      arrayElementType: { code: 'INT64' },
    });
  });

  it('adapts nested struct fields', () => {
    const client = {
      code: 'STRUCT',
      structType: {
        fields: [
          { name: 'n', type: { code: 'INT64' } },
          { name: 's', type: { code: 'STRING' } },
        ],
      },
    };
    assert.deepEqual(adaptClientType(client), {
      code: 'STRUCT',
      structType: {
        fields: [
          { name: 'n', type: { code: 'INT64' } },
          { name: 's', type: { code: 'STRING' } },
        ],
      },
    });
  });
});

describe('formatResultRow', () => {
  it('formats a row from native values', () => {
    const types = [
      { code: 'INT64' },
      { code: 'STRING' },
      {
        code: 'STRUCT',
        structType: {
          fields: [
            { name: 'n', type: { code: 'INT64' } },
            { name: 's', type: { code: 'STRING' } },
          ],
        },
      },
    ];
    const values = [42, null, { n: '7', s: 'x' }];
    const got = formatResultRow(types, values, simpleFormatConfig());
    assert.deepEqual(got, ['42', '<null>', '(7 AS n, x AS s)']);
  });
});

File: ruby/test/test_encoder.rb
# frozen_string_literal: true

require 'base64'
require 'json'
require 'minitest/autorun'
require 'pathname'

require_relative '../lib/spanvalue'

class EncoderTest < Minitest::Test
  CONFORMANCE_PATH = Pathname.new(__dir__).join('../../testdata/conformance.json').expand_path

  def self.conformance
    @conformance ||= JSON.parse(File.read(CONFORMANCE_PATH, encoding: 'UTF-8'))
  end

  def conformance
    self.class.conformance
  end

  def normalize_wire(value)
    return nil if value.nil? || Spanvalue::Proto.null_value?(value)

    case Spanvalue::Proto.value_kind(value)
    when 'bool' then Spanvalue::Proto.bool_value(value)
    when 'number' then Spanvalue::Proto.number_value(value)
    when 'string' then Spanvalue::Proto.string_value(value)
    when 'list' then Spanvalue::Proto.list_values(value).map { |elem| normalize_wire(elem) }
    else value
    end
  end

  def assert_wire_equal(expected, actual)
    exp = normalize_wire(expected)
    act = normalize_wire(actual)
    if exp.nil?
      assert_nil act
    else
      assert_equal exp, act
    end
  end

  def wire_to_native(typ, value)
    return nil if value.nil? || Spanvalue::Proto.null_value?(value)

    code = Spanvalue::Proto.type_code(typ)
    case code
    when Spanvalue::TypeCode::BOOL
      Spanvalue::Proto.bool_value(value)
    when Spanvalue::TypeCode::INT64, Spanvalue::TypeCode::ENUM
      Spanvalue::Proto.string_value(value)
    when Spanvalue::TypeCode::FLOAT32, Spanvalue::TypeCode::FLOAT64
      if Spanvalue::Proto.value_kind(value) == 'string'
        s = Spanvalue::Proto.string_value(value)
        return Float::NAN if s == 'NaN'
        return Float::INFINITY if s == 'Infinity'
        return -Float::INFINITY if s == '-Infinity'
      end
      Spanvalue::Proto.number_value(value)
    when Spanvalue::TypeCode::BYTES, Spanvalue::TypeCode::PROTO
      Base64.decode64(Spanvalue::Proto.string_value(value)).force_encoding(Encoding::ASCII_8BIT)
    when Spanvalue::TypeCode::STRING, Spanvalue::TypeCode::TIMESTAMP, Spanvalue::TypeCode::DATE, Spanvalue::TypeCode::NUMERIC,
         Spanvalue::TypeCode::INTERVAL, Spanvalue::TypeCode::UUID, Spanvalue::TypeCode::JSON
      Spanvalue::Proto.string_value(value)
    when Spanvalue::TypeCode::ARRAY
      elem_type = Spanvalue::Proto.array_element_type(typ)
      Spanvalue::Proto.list_values(value).map { |elem| wire_to_native(elem_type, elem) }
    when Spanvalue::TypeCode::STRUCT
      fields = Spanvalue::Proto.struct_fields(typ)
      Spanvalue::Proto.list_values(value).each_with_index.map do |elem, i|
        wire_to_native(Spanvalue::Proto.field_type(fields[i]), elem)
      end
    else
      value
    end
  end

  def test_encode_value_round_trip_conformance
    conformance.fetch('value_cases').each do |test_case|
      native = wire_to_native(test_case['type'], test_case['value'])
      encoded = Spanvalue.encode_value(test_case['type'], native)
      assert_wire_equal(test_case['value'], encoded)
      got = Spanvalue.format_value(test_case['type'], encoded, Spanvalue.simple_format_config)
      assert_equal test_case.dig('expected', 'simple'), got, "format after encode for #{test_case['name']}"
    end
  end

  def test_encode_struct_hash
    typ = {
      code: 'STRUCT',
      structType: {
        fields: [
          { name: 'n', type: { code: 'INT64' } },
          { name: 's', type: { code: 'STRING' } }
        ]
      }
    }
    encoded = Spanvalue.encode_value(typ, { 'n' => '1', 's' => 'foo' })
    assert_wire_equal(%w[1 foo], encoded)
    got = Spanvalue.format_value(typ, encoded, Spanvalue.literal_format_config)
    assert_equal 'STRUCT<n INT64, s STRING>(1, "foo")', got
  end

  def test_adapt_client_type
    client = {
      code: 'ARRAY',
      arrayElementType: { code: 'INT64' }
    }
    assert_equal(
      { code: 'ARRAY', arrayElementType: { code: 'INT64' } },
      Spanvalue::ClientTypeAdapter.adapt(client)
    )
  end

  def test_format_result_row
    types = [
      { code: 'INT64' },
      { code: 'STRING' },
      {
        code: 'STRUCT',
        structType: {
          fields: [
            { name: 'n', type: { code: 'INT64' } },
            { name: 's', type: { code: 'STRING' } }
          ]
        }
      }
    ]
    values = [42, nil, { 'n' => '7', 's' => 'x' }]
    got = Spanvalue.format_result_row(types, values, Spanvalue.simple_format_config)
    assert_equal ['42', '<null>', '(7 AS n, x AS s)'], got
  end
end
